Refactor the pipelines (batch and incremental) to have fixed workflows

chandrashekar-s opened the pull request:
## Description of what I changed
- Fixes #824
- Refactored the code such that multiple pipelines (one for each FHIR resource type) are created for each batch/incremental run.
- The multiple pipeline will be executed sequentially one after the other.
- Parallelism can be defined where in each parallel stream pipelines will be executed sequentially one at time.
- These changes will make sure that the memory required for Flink cluster is a constant factor and does not depend on the number of resource types.
- Regressions like Progress bar feature, pipeline metrics are addressed. 

## E2E test

Tested batch and incremental runs for multiple FHIR resource types keeping the Flink memory parameter `taskmanager.memory.network.max` constant and checked there is no impact on the number of resource types.

TESTED:

Given below are the stats (before and after the changes). Not much impact on the latencies, however after the changes the batch/incremental runs were successfully executed with less memory.

<google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #cccccc;}br {mso-data-placement:same-cell;}--></style>

Sl No | Type | # of FHIR types | # of Patients loaded | Time taken Before (sec) | Time taken Now (sec) | % Diff in time
-- | -- | -- | -- | -- | -- | --
1 | Batch | 3 | 3K | 120 | 55 | -54.17%
2 | Batch | 18 | 3K | 110 | 113 | 2.73%
3 | Batch | 18 | 20K | 1800 | 1840 | 2.22%
4 | Incremental | 3 | 50 (3K Base) | 51 | 50 | -1.96%
5 | Incremental | 18 | 50 (3K Base) | 119 | 106 | -10.92%
6 | Incremental | 18 | 100 (20K Base) | 350 | 430 | 22.86%

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [ ] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [ ] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

bashir2 commented:
Thanks @chandrashekar-s for this change; before I start the review, can you please take a look at the e2e test failure? It seems that at the batch step (with search API) it has completed the pipeline quickly but for some reasons was stuck (logs [here](https://console.cloud.google.com/cloud-build/builds/17b1a6dd-da9d-4566-9d0e-132229fa04ce;step=7?project=cloud-build-fhir)). I started a retry but I doubt it will fix it.

chandrashekar-s commented:
Thanks @bashir2 for pointing it out, I have fixed the issue and the build is successful now. Also currently the refactoring of pipelines is done to the JdbcHapi mode only and the rest of the modes are yet to be optimised.

bashir2 commented on pipelines/batch/src/main/java/com/google/fhir/analytics/ParquetMerger.java:
```diff
@@ -239,8 +245,20 @@ public static void main(String[] args) throws IOException {
 
-    Pipeline pipeline = createMergerPipeline(options, fhirContext);
-    EtlUtils.runMergerPipelineWithTimestamp(pipeline, options);
+    List<Pipeline> pipelines = createMergerPipelines(options, fhirContext);
```
nit: Please consider refactoring this logic into a function in `EtlUtil` as well; I think this is very similar to what you do in `FhirEtl.java` above. In the javadoc of that function please also describe briefly why we need to use a low number of pipelines running in parallel.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/CumulativeMetrics.java:
```diff
@@ -0,0 +1,40 @@
+ */
+package com.google.fhir.analytics.metrics;
+
+public final class CumulativeMetrics {
```
nit: Consider using `@Getter` of lombok to simplify this class.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkPipelineMetrics.java:
```diff
@@ -0,0 +1,173 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlinkPipelineMetrics implements PipelineMetrics {
```
Please add some javadoc for this class and in particular document its concurrency and singleton aspects.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkPipelineMetrics.java:
```diff
@@ -0,0 +1,173 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(FlinkPipelineMetrics.class);
+
+  private static ConcurrentHashMap<String, JobClient> jobClientMap = new ConcurrentHashMap<>();
```
I have not looked at all the details of Metrics related changes yet but is it fair to say that the new logic is basically very similar to the old one but instead of assuming that we only have one pipeline (and hence one "active" `jobClient`) we may have multiple now because there might be multiple pipelines?

Also the `static` requirements here are because of how `FlinkJobListener` is implemented correct? I wonder if it is possible to somehow set a new instance of  `FlinkPipelineMetrics` inside job-listener whenever we run a new batch of pipelines (basically I am trying to see if we can avoid this singleton pattern).

chandrashekar-s replied:
Yes, your understanding correct. The logic is similar to the older one, just that now we rely on the metrics of all pipelines submitted as per the current batch.

As per the implementation of the `FlinkJobListener` a new instance of it is created for every new pipeline, since for every batch we create multiple pipelines, setting one instance of `FlinkPipelineMetrics` for every instance of `FlinkJobListener` will not work.

chandrashekar-s replied:
Yes, having a static instance in the `PipelineMetricsProvider` provides the same desired effect. Hence the changes have been made accordingly. Also, the comments have been updated in the `FlinkPipelineMetrics` class.

bashir2 reviewed (commented):
Thanks @chandrashekar-s for the changes. I have not reviewed the whole PR yet but wanted to share the comments I have so far. I feel a little bit uneasy with the complexity that is needed for measuring progress but I don't have a better solution at the moment either.

chandrashekar-s commented on docker/config/application.yaml:
```diff
@@ -34,6 +34,7 @@ fhirdata:
   resourceList: "Patient,Encounter,Observation,Questionnaire"
   maxWorkers: 1
   numThreads: -1
+  maxParallelPipelineStreams: 2
```
I was using this parameter over [here](https://github.com/google/fhir-data-pipes/pull/840/commits/4444891b9ca52324c25564f774a30d1b7900d56e#diff-67c72b661b13c2890d4afdb112f4ebe3e8da96902491df4a98066730f6083a8dR169), but now I have eliminated this parameter and using a [constant number 2](https://github.com/google/fhir-data-pipes/pull/840/commits/34794a16de2c2a811bf9bd3f95bb8b61f413d5e6#diff-67c72b661b13c2890d4afdb112f4ebe3e8da96902491df4a98066730f6083a8dR170).

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -46,46 +52,89 @@ static void logMetrics(MetricResults metricResults) {
-   * Runs the given `pipeline` and if the output is to a Parquet DWH, also writes a timestamp file
-   * indicating when the pipeline was started. This is useful for future incremental runs.
+   * Runs the given `pipelines` and if the output is to a Parquet DWH, also writes a timestamp file
+   * indicating when the pipelines were started. This is useful for future incremental runs.
```
Done

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtl.java:
```diff
@@ -398,9 +418,17 @@ public static void main(String[] args) throws PropertyVetoException, IOException
+    if (pipelines != null && !pipelines.isEmpty()) {
+      ExecutorService executor = null;
+      try {
+        executor = Executors.newFixedThreadPool(2);
```
`maxParallelPipelineStreams` has been removed now and this code is refactored to use the constant `EtlUtils.NO_OF_PARALLEL_PIPELINES`

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtl.java:
```diff
@@ -328,15 +341,22 @@ public void processElement(ProcessContext context) {
+          PipelineMetricsFactory.getPipelineMetrics(options.getRunner());
+      if (pipelineMetrics != null) {
+        pipelineMetrics.clearAllMetrics();
+        pipelineMetrics.setTotalNoOfResources(totalNoOfResources);
```
Currently the progress bar is not implemented for other modes, it only works for `Jdbc` mode.

chandrashekar-s commented:
Thanks @bashir2 for the review. I have addressed the review comments, please have a look. Also, I checked if the refactoring is needed for the `FHIR Search` mode, it turns out that this mode has no `reshuffling's` and hence the number of network buffers used is very low. However, when the `activePeriod` flag is set then for this function `Sum.integersPerKey()` over [here](https://github.com/google/fhir-data-pipes/blob/a4f90baee30482e95d8b4253d45d993cc0c8ec6e/pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtl.java#L114) it allocates N^2 number of  network buffers (N=No of Cores). So I feel no refactoring is required for `FHIR Search` mode as we have already allocated enough buffers to handle 9 * N^2 buffers (needed for merge pipeline).

chandrashekar-s commented on docker/config/flink-conf.yaml:
```diff
@@ -16,7 +16,7 @@
 # This is needed to prevent an "Insufficient number of network buffers"
 # exceptions when running the merger on large input with many workers.
-taskmanager.memory.network.max: 5gb
+taskmanager.memory.network.max: 1gb
```
This is in progress, will update once done.

bashir2 commented on pipelines/controller/src/main/java/com/google/fhir/analytics/metrics/Stats.java:
```diff
@@ -42,33 +38,19 @@ public class Stats {
     Stats stats = new Stats();
-    long totalNoOfResources = 0l;
-    for (MetricResult<GaugeResult> gauge : metricQueryResults.getGauges()) {
-      if (gauge.getName().getName().startsWith(MetricsConstants.TOTAL_NO_OF_RESOURCES)) {
```
So it seems these `TOTAL_NO_OF_RESOURCES` metrics are not needed anymore and I wonder do we still need the  `DoFn` in `FhirEtl` that were just to register these metrics? Those are at line 309 of `FhirEtl.java` above (of the old code, or line 320 of the new one). It would be great if we can remove those `ParDo`s (if you remember, we had some discussions around those when you originally added them).

chandrashekar-s replied:
Yes, now the `TOTAL_NO_OF_RESOURCES` are not needed for tracking the progress bar. However, these metrics which gets reported are being in the UI. Probably, we will have to report these if needs to be shown in the UI. Currently, thinking of ways to report this.

bashir2 replied:
I see, but given that those stats are somewhat broken now (because of #839), I would suggest to remove those `ParDo`s in this PR and implement a better way of reporting total numbers as part of fixing #839.

chandrashekar-s replied:
Ok, removed the ParDo function now, will address it as part of https://github.com/google/fhir-data-pipes/issues/839

bashir2 reviewed (commented):
Thanks for the changes @chandrashekar-s ; I have made some suggestions around how metrics are collected. I will look at the details of the progress-calculation logic after those are resolved.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -46,46 +52,89 @@ static void logMetrics(MetricResults metricResults) {
+      }
+    } catch (InterruptedException | ExecutionException e) {
+      log.error("Error while executing the pipelines", e);
+      throw new RuntimeException(e);
```
Done

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -30,10 +33,18 @@
+   * {@value #NO_OF_PARALLEL_PIPELINES} Total number of pipelines that can be executed in parallel
+   * at a given point in time.
+   */
+  public static final Integer NO_OF_PARALLEL_PIPELINES = 2;
```
Done

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -17,17 +17,34 @@
 class EtlUtils {
+
+  /**
+   * {@value #NO_OF_PARALLEL_PIPELINES} Total number of pipelines that can be executed in parallel
```
Done

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkPipelineMetrics.java:
```diff
@@ -0,0 +1,220 @@
+   * @param totalNoOfResources the total number of resources to be processed in the current batch
+   */
+  @Override
+  public synchronized void setTotalNoOfResources(long totalNoOfResources) {
```
Probably this might give a better clarity that this tracks the total number of all resources in the batch rather the resources in one pipeline (debatable though).

bashir2 replied:
Sure; maybe only do the `No` -> `Num` suggestion then.

chandrashekar-s replied:
Done

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -46,46 +63,144 @@ static void logMetrics(MetricResults metricResults) {
-  static PipelineResult runPipelineWithTimestamp(Pipeline pipeline, FhirEtlOptions options)
-      throws IOException {
+  static List<PipelineResult> runMultiplePipelinesWithTimestamp(
+      List<Pipeline> pipelines, FhirEtlOptions options, Executor executor) throws IOException {
```
As discussed, I agree with you that the `Executor` need not be exposed in the non-private methods, since we initialise the pool with a limited number of threads and is not much of a overhead to recreate them each time.

chandrashekar-s commented on pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java:
```diff
@@ -160,6 +164,11 @@ private void initDwhStatus() {
+    Preconditions.checkArgument(dataProperties.getMaxWorkers() > 0, "maxWorkers should be > 0");
+    executor =
+        Executors.newFixedThreadPool(
+            dataProperties.getMaxWorkers() * EtlUtils.NO_OF_PARALLEL_PIPELINES);
```
As commented in the previous thread, the executor has been removed and is left as an implementation detail in `EtlUtils`. So removed these changes over here.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/FlinkJobListener.java:
```diff
@@ -34,7 +34,7 @@ public void onJobSubmitted(@Nullable JobClient jobClient, @Nullable Throwable th
 
     logger.info("Submitting the job with ID {} ", this);
-    FlinkPipelineMetrics.setJobClient(jobClient);
+    FlinkPipelineMetrics.addJobClient(jobClient);
```
Yes, have made the corresponding changes.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/metrics/PipelineMetricsFactory.java:
```diff
@@ -19,16 +19,13 @@
-import org.springframework.stereotype.Component;
 
-@Component
 public class PipelineMetricsFactory {
```
Ok, done.

chandrashekar-s commented:
@bashir2 thanks for the review. I have addressed most of the comments in the latest commit. Please have a look.

bashir2 reviewed (approved):
Thanks @chandrashekar-s for the changes. Please feel free to merge this after addressing the remaining minor comments.

chandrashekar-s commented on pipelines/batch/src/main/java/com/google/fhir/analytics/EtlUtils.java:
```diff
@@ -46,46 +60,138 @@ static void logMetrics(MetricResults metricResults) {
+   * @return the list of pipeline results
+   * @throws IOException
+   */
+  static List<PipelineResult> runMultiplePipelines(
```
Yes, removed this function now.
